### ISSUE TITLE
Use host root as --template/reproduce `firejail --overlay` functionality using systemd-nspawn

### ISSUE THREAD

**Issue by oxwivi** [systemd/systemd#9044]
The `--overlay` flag for `firejail` makes COW namespace of the host system, allowing the user to install, configure and generally fuck with the system in the namespace without actually affecting the base root directory. It is incredibly useful to quickly test something without having to spin a complete new container, and a completely different environment than the host.

Unfortunately, as I reported [here](
https://github.com/netblue30/firejail/issues/1908), any `systemd` services installed in the `firejail` overlay namespace can't actually work because it doesn't have access to the `systemd` namespace as [explained](
https://github.com/netblue30/firejail/issues/1908#issuecomment-383924855) by `firejail`'s developer.

Because of that, I wanted to try my hand at doing the same using `systemd-nspawn`, but haven't met much success:
```
$ sudo systemd-nspawn --template=/ -D test -U --bind=/user/username/Downloads
Couldn't create snapshot /home/ontoki/test from /: Invalid cross-device link
```
`sudo systemd-nspawn -UD / -x` successfully spawns an ephemeral container, but network doesn't work and I need to use `-b` to boot before it does.

Is there anyway the `firejail --overlay --no-profile` functionalities can be repoduced using `systemd-nspawn`?

**Comment by boucman** [comment]
Looking at the explanation provided by the firejail maintainer, I don't really see how to work around that limitation

* The service is installed in the overlay (which is the whole point of firejail)
* systemd itself is in the root namespace "below" the overlay

you don't want systemd to see what's in the overlay, that would break the idea of firejail isolating your application from the host, so i don't really know what can be done...

to put it differently, having systemd know about firejailed services would break the firejail... so i'm not sure if it's doable or desirable to do that...

**Comment by oxwivi** [comment]
I'm well aware it'd break the sandbox, but as I said in my last comment there, security is not my goal here, but simply isolating the processes and filesystem so that I can quickly create and get rid of the "sandboxed" environment that I created using `firejail`.

**Comment by poettering** [comment]
Can you explain what precisely you are asking for? I don't know firejail, hence please describe the low-level operations you are looking for

**Comment by oxwivi** [comment]
Put simply, I want to use btrfs snapshot of my host system as the root of the container, so whatever I do in the container will feel like I'm doing it on the host system, but, of course it's only happening in the container, so I can get rid of it anytime. btrfs COW snapshot for the root, and the container itself to isolate processes and services running within it.

Though the `firejail` capability is a bit more flexible with the `overlay` flag, it can still see the changes I make outside `firejail`'s isolation, but whatever changes made within `firejail` remains in it.

Kind of like the sandboxed environment `snap`, `flatpak` and friends provide, I suppose. They can see the underlying host, but any changes they make are isolated from the host.

**Comment by poettering** [comment]
There's systemd-nspawn `--ephemeral`, which relies on reflinks to be efficient. Implementing this via overlayfs has been requested as part of #3847. 

Hence closing as duplicate.
